### DISCUSSION TITLE
Mapping Enum variants with same enum keys

### DISCUSSION THREAD

**Discussion by aki-ks** [nartc/mapper#393] [Q&A] [answered]
### Is your feature request related to a problem? Please describe.

When I'm mapping between two kind of datastructures, I'm often facing the problem that they're both using equivalent enums but these encode the values as different constants or have different variant names.

```typescript
class UserEntity {
  @AutoMap()
  name: string;

  @AutoMap()
  favoriteColor: ColorEntity;
}

enum ColorEntity {
  Red = 'r',
  Green = 'g',
  Blue = 'b',
  Other = 'o',
}
```

```typescript
class UserDto {
  @AutoMap()
  name: string;

  @AutoMap()
  favoriteColor: ColorDto;
}

enum ColorDto {
  Red = 'red',
  Green = 'green',
  Blue = 'blue',
  Custom = 'custom',
}
```

How can I teach AutoMapper to map `ColorEntity` to `ColorDto`?

### Describe the solution you'd like

I'd love to see automapper handle enums like any other class, so a mapping between two enums can be defined. The variants of the enum are mapped like typescript fields:

- When both enums have variants of the same name, then they are mapped to each other (`Red`, `Green`, `Blue` in the upper example).
- Variants that the two enum types do not have in common require a manual mapping (`Custom`/`Other` in the example)

```typescript
mapper.createMap(ColorEntity, ColorDto)
  .forMember(dto => dto.Custom, mapFrom(entity => entity.Other))
```

### Additional context

I barely know typescript reflection, so I can't judge whether this idea is realizable.
Can the variants of enums be discovered via reflection or whatever is required for this feature?

**Comment by nartc** [comment, marked as answer]
Thank you for showing interest and requesting this feature.

However, Enums are complicated: string enums, number enums, mixed enums, and const enums... Hence, it is tricky to get some generic support for enum in this sense.

What I can suggest you to do is to create a `Resolver` and use a `switch` statement

```ts
export const colorResolver: Resolver<
  {color: ColorEntity},
  {color: ColorDto},
  ColorDto
> = {
  resolve(source: TypeConverterSource): ColorDto {
    switch (source.color) {
      case Color.Red:
        return ColorDto.Red;
      case Color.Green:
        return ColorDto.Green;
      case Color.Blue:
        return ColorDto.Blue;
    }
  },
};
```

then use it on a `mapFrom()`

```ts
mapper.createMap(Source, Destination)
   .forMember(d => d.color, mapFrom(colorResolver))
```

**Comment by nartc** [comment]
@aki-ks can we close this issue?

**Comment by aki-ks** [comment]
If you want to leave it at your suggestion and don't plan to add extended support for enums to the library, then sure.
Your recommendation has already helped me a lot. :+1:

**Comment by nartc** [comment]
Thank you. I'll convert this to a discussion then :)
